server/src/domain/storage/storage.core.ts

```diff
@@ -143,26 +168,93 @@ export class StorageCore {
       this.logger.log(`Attempting to finish incomplete move: ${move.oldPath} => ${move.newPath}`);
       const oldPathExists = await this.repository.checkFileExists(move.oldPath);
       const newPathExists = await this.repository.checkFileExists(move.newPath);
-      const actualPath = newPathExists ? move.newPath : oldPathExists ? move.oldPath : null;
+      const actualPath = oldPathExists ? move.oldPath : newPathExists ? move.newPath : null;
       if (!actualPath) {
         this.logger.warn('Unable to complete move. File does not exist at either location.');
         return;
       }
 
-      this.logger.log(`Found file at ${actualPath === move.oldPath ? 'old' : 'new'} location`);
+      const fileAtNewLocation = actualPath === move.newPath;
+      this.logger.log(`Found file at ${fileAtNewLocation ? 'new' : 'old'} location`);
+
+      if (fileAtNewLocation) {
+        if (!(await this.verifyNewPathContentsMatchesExpected(move.oldPath, move.newPath, assetInfo))) {
+          this.logger.fatal(
+            `Skipping move as file verification failed, old file is missing and new file is different to what was expected`,
+          );
+          return;
+        }
+      }
 
       move = await this.moveRepository.update({ id: move.id, oldPath: actualPath, newPath });
     } else {
       move = await this.moveRepository.create({ entityId, pathType, oldPath, newPath });
     }
 
+    if (pathType === AssetPathType.ORIGINAL && !assetInfo) {
+      this.logger.warn(`Unable to complete move. Missing asset info for ${entityId}`);
+      return;
+    }
+
     if (move.oldPath !== newPath) {
-      await this.repository.moveFile(move.oldPath, newPath);
+      try {
+        this.logger.debug(`Attempting to rename file: ${move.oldPath} => ${newPath}`);
+        await this.repository.rename(move.oldPath, newPath);
+      } catch (err: any) {
+        if (err.code !== 'EXDEV') {
+          this.logger.warn(
+            `Unable to complete move. Error renaming file with code ${err.code} and message: ${err.message}`,
+          );
+          return;
+        }
+        this.logger.debug(`Unable to rename file. Falling back to copy, verify and delete`);
+        await this.repository.copyFile(move.oldPath, newPath);
+
+        if (!(await this.verifyNewPathContentsMatchesExpected(move.oldPath, newPath, assetInfo))) {
+          this.logger.warn(`Skipping move due to file size mismatch`);
+          return;
+        }
```

Comment:
In this case the asset still uses the old path, right? Should we do any cleanup of the new one?

Reply:
I think that's a good call because otherwise 180 would still be true in subsequent tries and we'd be running in 181/182. Thus, the file never will be fixed.

Reply:
Yea we could attempt a cleanup, I just wanted to avoid doing even more stuff in here tbh

Reply:
I added a cleanup but if it fails we just error out 😄